activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -115,7 +115,16 @@ def load_records(&block)
             # Make several smaller queries if necessary or make one query if the adapter supports it
             slices = owner_keys.each_slice(klass.connection.in_clause_length || owner_keys.size)
             @preloaded_records = slices.flat_map do |slice|
-              records_for(slice).load(&block)
+              records = records_for(slice)
```

Comment:
It does make more sense to move that logic to `records_for` that knows how the records are instantiated and can give them the `block` as early as possible in their instantiation. If the other comment makes sense as well, I'll change the PR.